File: vscode-extension/src/LocalIdentifierManager.ts
import * as os from 'os';
import * as crypto from 'crypto';
import * as vscode from 'vscode';

/**
 * Local identifier manager
 * Responsible for generating and managing local unique identifiers for use by various components
 * Supports multi-project instances, distinguishes different IDEA project windows by project path
 * Supports VSCode multi-version, distinguishes different processes of the same project by PID
 */
export class LocalIdentifierManager {
    private static _instance: LocalIdentifierManager | null = null;
    private readonly _identifier: string;

    private constructor() {
        this._identifier = this.generateLocalIdentifier();
    }

    /**
     * Get singleton instance
     */
    static getInstance(): LocalIdentifierManager {
        if (!LocalIdentifierManager._instance) {
            LocalIdentifierManager._instance = new LocalIdentifierManager();
        }
        return LocalIdentifierManager._instance;
    }

    /**
     * Get local unique identifier
     */
    get identifier(): string {
        return this._identifier;
    }

    /**
     * Generate local unique identifier
     * Format: hostname-projectHash-pid
     * - projectHash: Solves the problem of same PID for IDEA multi-project windows
     * - pid: Solves the problem of different PIDs for VSCode multi-version same project
     */
    private generateLocalIdentifier(): string {
        try {
            const hostname = os.hostname();
            const projectHash = this.generateProjectHash();
            const pid = process.pid;
            return `${hostname}-${projectHash}-${pid}`;
        } catch (e) {
            return `unknown-${Date.now()}-${Math.floor(Math.random() * 10000)}`;
        }
    }

    /**
     * Generate project hash
     * Generate short hash based on project path, used to distinguish different project instances
     */
    private generateProjectHash(): string {
        try {
            // Get VSCode workspace path
            const workspaceFolders = vscode.workspace.workspaceFolders;
            const projectPath = workspaceFolders && workspaceFolders.length > 0
                ? workspaceFolders[0].uri.fsPath
                : 'unknown-project';

            // Generate MD5 hash
            const hash = crypto.createHash('md5');
            hash.update(projectPath);
            const hashBytes = hash.digest();

            // Take first 3 bytes (6 hexadecimal characters) as project hash
            return hashBytes.subarray(0, 3).toString('hex');
        } catch (e) {
            // If hash generation fails, use last 6 digits of timestamp as fallback
            return Date.now().toString().slice(-6);
        }
    }
}

File: jetbrains-plugin/src/main/kotlin/com/vscode/jetbrainssync/LocalIdentifierManager.kt
package com.vscode.jetbrainssync

import com.intellij.openapi.project.Project
import java.net.InetAddress
import java.security.MessageDigest
import java.util.concurrent.atomic.AtomicLong

/**
 * Local identifier manager
 * Responsible for generating and managing local unique identifiers for use by various components
 * Supports multi-project instances, distinguishing different IDEA project windows through project paths
 * Supports VSCode multi-version, distinguishing different processes of the same project through PID
 */
class LocalIdentifierManager(private val project: Project) {

    /**
     * Local unique identifier
     * Generated on first access, format: hostname-projectHash-pid
     */
    val identifier: String by lazy { generateLocalIdentifier() }

    /**
     * Project-specific message sequence number generator
     * Solves the problem of multi-project instances sharing global sequence numbers
     */
    private val messageSequence = AtomicLong(0)

    /**
     * Generate local unique identifier
     * Format: hostname-projectHash-pid
     * - projectHash: Solves the problem of IDEA multi-project windows having the same PID
     * - pid: Solves the problem of VSCode multi-version having different PIDs for the same project
     */
    private fun generateLocalIdentifier(): String {
        return try {
            val hostname = InetAddress.getLocalHost().hostName
            val projectHash = generateProjectHash()
            val pid = ProcessHandle.current().pid()
            "$hostname-$projectHash-$pid"
        } catch (e: Exception) {
            "unknown-${System.currentTimeMillis()}-${(Math.random() * 10000).toInt()}"
        }
    }

    /**
     * Generate project hash value
     * Generate short hash based on project path, used to distinguish different project instances
     */
    private fun generateProjectHash(): String {
        return try {
            val path = project.basePath ?: "unknown-project"
            val md = MessageDigest.getInstance("MD5")
            val hashBytes = md.digest(path.toByteArray())
            // Take first 3 bytes (6 hexadecimal characters) as project hash
            hashBytes.take(3).joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            // If hash generation fails, use last 6 digits of timestamp as fallback
            System.currentTimeMillis().toString().takeLast(6)
        }
    }

    /**
     * Generate project-specific message ID
     * Format: {localIdentifier}-{sequence}-{timestamp}
     * Each project instance has independent sequence number, avoiding multi-instance conflicts
     */
    fun generateMessageId(): String {
        val sequence = messageSequence.incrementAndGet()
        val timestamp = System.currentTimeMillis()
        return "$identifier-$sequence-$timestamp"
    }

}